File: cpp/teach_DIP.cpp
#include <iostream>
#include <string>
#include <chrono>
#include <thread>
#include <random>

using namespace std::chrono;

// Dependency Inversion Principle - DIP
// SuperModules and SubModule depend on interfaces, but not between themselves

//
//  Logger Interface
//
class ILogger{
public:
    virtual ~ILogger(){}

    virtual void log(const std::string& message) = 0;
};

//
//  Proto of Separated Logger Module
//
class Logger: public ILogger{
public:
    explicit Logger(){}

    void log(const std::string& message){
        std::cout << "Just logging " << message << " ..." << std::endl;
    }
};

//
// Proto of DataBase Module
//
class DataBase: public ILogger{
public:
    explicit DataBase(std::string name):
        m_name(name), m_isReady(false)
    {}

    std::string getName() const{
        return m_name;
    }

    void log(const std::string& message){
        std::cout << "<DB> Logging " << message << " ..." << std::endl;
    } 

    void open(){
        m_isReady = true;
        log("<DB> " + m_name + " opened for transactions!");
    }

    void close(){
        m_isReady = false;
        log("<DB> " + m_name + " closed for transactions!");
    }

    void do_request(std::string request){
        if (!m_isReady){
            log("<DB> " + m_name + " is not ready for transactions!");
            return;
        }

        std::this_thread::sleep_for(300ms);
        log("<DB> " + m_name + " query '" + request + "' completed!");
    }

private:
    std::string m_name;
    bool        m_isReady;
};


//
//  Proto of File Module
//
class File: public ILogger{
public:
    explicit File(std::string path):
        m_path(path),
        m_isReady(false)
    {}

    std::string getPath() const{
        return m_path;
    }

    void log(const std::string& message){
        std::cout << "<FILE> Logging " << message << " ..." << std::endl;
    }

    void open(){
        m_isReady = true;
        log("<FILE> " + m_path + " opened for operatoins!");
    }

    void close(){
        m_isReady = false;
        log("<FILE> " + m_path + " closed for operations!");
    }

    std::string read(){
        if (!m_isReady){
            log("<FILE> " + m_path + " is not ready for operations!");
            return std::string();
        }

        std::this_thread::sleep_for(10ms);
        log("<FILE> " + m_path + " has been read!");
        return "string";
    }

    void write(std::string s){
        if(!m_isReady){
            log("<FILE> " + m_path + " is not ready for operations!");
            return;
        }

        std::this_thread::sleep_for(10ms);
        log("<FILE> " + m_path + " wrote string " + s + "!");
    }

    void clear(){
        if(!m_isReady){
            log("<FILE> " + m_path + " is not ready for operations!");
            return;
        }
        
        std::mt19937 mt{std::random_device{}()};
        std::uniform_real_distribution<> rand_range{0.01, 2.5};
        double rand_val = rand_range(mt);
        std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(rand_val * 1000)));

        log("<FILE> " + m_path + " was cleared!");
    }

private:
    std::string m_path;
    bool        m_isReady;
};


int main(){
    Logger logger;
    logger.log("bumbaracha!");

    DataBase db("Sample");
    db.open();
    db.do_request("create table init");
    db.close();

    File f("1.docx");
    f.open();
    std::string stroke = f.read();
    std::cout << stroke << std::endl;
    f.clear();
    f.write(stroke);
    f.close();

    std::cout << "Done!" << std::endl;
    
    return 0;
}

File: python/teach_DIP.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import time
import random


# Dependency Inversion Principle - DIP
# Super Modules and Sub Modules depend on interfaces, but not between themselves


from abc import ABC, abstractmethod


class ILogger(ABC):
    """
    Interface of Logger
    """
    @abstractmethod
    def log(self, message: str):
        pass


class Logger(ILogger):
    """Proto of Separated Logger Module"""
    def __init__(self):
        pass

    def log(self, message: str):
        print(f"Just logging {message} ...")


class DataBase(ILogger):
    """
    Proto of DataBase Module
    """
    def __init__(self, name: str):
        self.__name = name
        self.__isReady = False

    @property
    def name(self):
        return self.__name

    def log(self, message: str):
        print(f"<DB> Logging {message} ...")

    def open(self):
        self.__isReady = True
        self.log(f"<DB> {self.__name} opened for transactions!")

    def close(self):
        self.__isReady = False
        self.log(f"<DB> {self.__name} closed for transactions!")

    def do_request(self, request: str):
        if not self.__isReady:
            self.log(f"<DB> {self.__name} is not ready for transactions!")
            return

        time.sleep(0.3)
        self.log(f"<DB> {self.__name} query '{request}' completed!")


class File(ILogger):
    """
    Proto of File Module
    """
    def __init__(self, path: str):
        self.__path = path
        self.__isReady = False

    @property
    def path(self):
        return self.__path

    def log(self, message: str):
        print(f"<FILE> Logging {message} ...")

    def open(self):
        self.__isReady = True
        self.log(f"<FILE> {self.__path} opened for operations!")

    def close(self):
        self.__isReady = False
        self.log(f"<FILE> {self.__path} closed for operations!")

    def read(self) -> str:
        if not self.__isReady:
            self.log(f"<FILE> {self.__path} is not ready for operations!")
            return str()

        time.sleep(0.01)
        self.log(f"<FILE> {self.__path} has been read!")
        return "string"

    def write(self, s: str):
        if not self.__isReady:
            self.log(f"<FILE> {self.__path} is not ready for operations!")
            return

        time.sleep(0.01)
        self.log(f"<FILE> {self.__path} wrote string {s}!")

    def clear(self):
        if not self.__isReady:
            self.log(f"<FILE> {self.__path} is not ready for operations!")
            return

        time.sleep(random.uniform(0.01, 2.5))
        self.log(f"<FILE> {self.__path} was cleared!")


if __name__ == "__main__":
    logger = Logger()
    logger.log("bumbaracha!")

    db = DataBase("Sample")
    db.open()
    db.do_request("create table init")
    db.close()

    f = File("1.docx")
    f.open()
    stroke = f.read()
    print(stroke)
    f.clear()
    f.write(stroke)
    f.close()

    print("Done!")

